std/array_list.zig

```diff
@@ -193,27 +194,105 @@ pub fn AlignedArrayList(comptime T: type, comptime A: u29) type {
             return self.pop();
         }
 
+        // an iterator which allows you to remove the current element
+        // while iterating through it.
         pub const Iterator = struct {
-            list: *const Self,
-            // how many items have we returned
-            count: usize,
+            list: *Self,
+
+            // index of the next element to return
+            next_index: usize = 0,
+
+            // the last index we removed
+            removed_index: ?usize = null,
 
             pub fn next(it: *Iterator) ?T {
-                if (it.count >= it.list.len) return null;
-                const val = it.list.at(it.count);
-                it.count += 1;
+                if (it.next_index >= it.list.len) return null;
+                const val = it.list.at(it.next_index);
+
+                it.next_index += 1;
+
                 return val;
             }
 
             pub fn reset(it: *Iterator) void {
-                it.count = 0;
+                it.next_index = 0;
+                it.removed_index = null;
+            }
+
+            //
+            // NOTE: inlining these removal routines is important.
+            // Allows them to get with 1% of the performance of ArrayList.swapRemove/orderedRemove.
+            //
+
+            pub inline fn swapRemove(it: *Iterator) T {
+                const removed_already = blk: {
```

Comment:
could this be an expression? `if (it.removed_index) |idx| idx==it.next_index else false`

Reply:
Alternatively move it directly into the `if` below: `if (it.removed_index == null or it.removed_index.? != it.next_index) {`

Reply:
I thought `or` evaluated both expressions?

Reply:
Changed to use this. Though, I'm not sure why it works. Does Zig specifically check for this pattern of `if (opt == null or opt.? == value)` ?
Only I'd expect this to evaluate both expressions and crash.

Reply:
I'd be very surprised if zig didn't do [Short circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation#Support_in_common_programming_languages)

Reply:
Easy to test:

```zig
if (true or unreachable) std.debug.warn("hi");
```

(Spoiler: zig does short circuit evaluation)